Running Test Classes with Inherited @Factory and @DataProvider Annotated Non-Static Methods Fail

### TestNG Version

7.6.1

### Expected behavior

Tests with inherited @Factory and @DataProvider annotated methods (non-static) from an abstract class fail while trying to instantiate the DataProvider.  

```
public class InheritedTestFactoryAndDataProvider
{
	public static abstract class AbstractTestClassGenerator
	{
		@DataProvider(name="dataProvider")
		public Object[] dataProvider()
		{
			return new String[]{ "foo", "bar" };
		}
		
		@Factory(dataProvider="dataProvider")
		public Object[] testFactory(String value)
		{
			return new Object[] { new TestClass() };
		}
	}
	
	public static class TestClassGenerator extends AbstractTestClassGenerator
	{
		// assume other configuration etc. is happening here
	}
	
	public static class TestClass 
	{
		@Test
		public void test()
		{
		}
	}
	
}
```


### Actual behavior

AFAICS the proceeding should work according to the docs/javadoc etc. 

However this produces the following error:

```
Cannot instantiate class InheritedTestFactoryAndDataProvider$AbstractTestClassGenerator
	at org.testng.internal.objects.InstanceCreator.newInstance(InstanceCreator.java:41)
	at org.testng.internal.objects.InstanceCreator.newInstance(InstanceCreator.java:59)
	at org.testng.ITestObjectFactory.newInstance(ITestObjectFactory.java:10)
	at org.testng.internal.objects.SimpleObjectDispenser.dispense(SimpleObjectDispenser.java:60)
	at org.testng.internal.objects.GuiceBasedObjectDispenser.dispense(GuiceBasedObjectDispenser.java:28)
	at org.testng.internal.Parameters.findDataProvider(Parameters.java:626)
	at org.testng.internal.Parameters.findDataProvider(Parameters.java:510)
	at org.testng.internal.Parameters.handleParameters(Parameters.java:763)
	at org.testng.internal.FactoryMethod.invoke(FactoryMethod.java:160)
	at org.testng.internal.TestNGClassFinder.processFactory(TestNGClassFinder.java:174)
	at org.testng.internal.TestNGClassFinder.processMethod(TestNGClassFinder.java:138)
	at org.testng.internal.TestNGClassFinder.processClass(TestNGClassFinder.java:129)
	at org.testng.internal.TestNGClassFinder.<init>(TestNGClassFinder.java:67)
	at org.testng.TestRunner.initMethods(TestRunner.java:441)
	at org.testng.TestRunner.init(TestRunner.java:335)
	at org.testng.TestRunner.init(TestRunner.java:288)
	at org.testng.TestRunner.<init>(TestRunner.java:218)
	at org.testng.remote.support.RemoteTestNG6_12$1.newTestRunner(RemoteTestNG6_12.java:33)
	at org.testng.remote.support.RemoteTestNG6_12$DelegatingTestRunnerFactory.newTestRunner(RemoteTestNG6_12.java:66)
	at org.testng.ITestRunnerFactory.newTestRunner(ITestRunnerFactory.java:52)
	at org.testng.SuiteRunner$ProxyTestRunnerFactory.newTestRunner(SuiteRunner.java:706)
	at org.testng.SuiteRunner.init(SuiteRunner.java:225)
	at org.testng.SuiteRunner.<init>(SuiteRunner.java:115)
	at org.testng.TestNG.createSuiteRunner(TestNG.java:1349)
	at org.testng.TestNG.createSuiteRunners(TestNG.java:1325)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1167)
	at org.testng.TestNG.runSuites(TestNG.java:1099)
	at org.testng.TestNG.run(TestNG.java:1067)
	at org.testng.remote.AbstractRemoteTestNG.run(AbstractRemoteTestNG.java:115)
	at org.testng.remote.RemoteTestNG.initAndRun(RemoteTestNG.java:251)
	at org.testng.remote.RemoteTestNG.main(RemoteTestNG.java:77)
Caused by: java.lang.InstantiationException
	at java.base/jdk.internal.reflect.InstantiationExceptionConstructorAccessorImpl.newInstance(InstantiationExceptionConstructorAccessorImpl.java:48)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
	at org.testng.internal.objects.InstanceCreator.newInstance(InstanceCreator.java:38)
	... 30 more
```

### Is the issue reproducible on runner?

- [ ] Shell
- [ ] Maven
- [x] Gradle
- [ ] Ant
- [x] Eclipse
- [ ] IntelliJ
- [ ] NetBeans


### Test case sample

See above.

### Thoughts on Issue

Debugging the code it seems that there is an issue where the @DataProvider annotation is processed in JDK15TagFactory which always winds up setting the dataProviderClass member to a non-null value even if the dataProviderClass annotation attribute is not set.  The non-null value in this case happens to be the abstract class where the annotation is read.   Later on this value gets passed to org.testng.internal.Parameters.findDataProvider as seen in the stack above.  This method assumes that dataProviderClass should be null if the dataProviderClass member is not set (at least that's my interpretation given the doc on the annotation and the context of the code) and therefore sets the shouldBeStatic flag, this then forces an attempt to re-instantiate the class, instead of just using the already instantiated instance of the factory class, which is the else case e.g. 

```
        if (shouldBeStatic && (m.getModifiers() & Modifier.STATIC) == 0) {
          IObjectDispenser dispenser = Dispenser.newInstance(objectFactory);
          BasicAttributes basic = new BasicAttributes(clazz, dataProviderClass);
          CreationAttributes attributes = new CreationAttributes(context, basic, null);
          instanceToUse = dispenser.dispense(attributes);
        } else {
          instanceToUse = instance;
        }
```

Unfortunately, I'm not sure what the actual solution is, since I may be missing some context around the Guice implementation stuff, but at least in this case the logic around setting shouldBeStatic seems wrong, and causes the test case above to fail.

### Contribution guidelines

Incase you plan to raise a pull request to fix this issue, please make sure you refer our [Contributing](.github/CONTRIBUTING.md) section for detailed set of steps.

--- a/CHANGES.txt
+++ b/CHANGES.txt
@@ -1,4 +1,5 @@
 Current
+Fixed: GITHUB-2800: Running Test Classes with Inherited @Factory and @DataProvider Annotated Non-Static Methods Fail (Krishnan Mahadevan)
 New: Ability to provide custom error message for assertThrows\expectThrows methods (Anatolii Yuzhakov)
 Fixed: GITHUB-2780: Use SpotBugs instead of abandoned FindBugs
 Fixed: GITHUB-2801: JUnitReportReporter is too slow

--- a/gradle.properties
+++ b/gradle.properties
@@ -5,7 +5,7 @@ kotlin.code.style=official
 # Note: testng.kotlin-library.gradle.kts adds kotlin-stdlib for testImplementation
 kotlin.stdlib.default.dependency=false
 
-testng.version=7.6.1
+testng.version=7.6.2
 
 group=org.testng
 

--- a/testng-core/src/main/java/org/testng/internal/Parameters.java
+++ b/testng-core/src/main/java/org/testng/internal/Parameters.java
@@ -610,7 +610,8 @@ public class Parameters {
 
     Class<?> cls = clazz.getRealClass();
     boolean shouldBeStatic = false;
-    if (dataProviderClass != null) {
+    boolean isDataProviderClassAbstract = Modifier.isAbstract(cls.getModifiers());
+    if (dataProviderClass != null && !isDataProviderClassAbstract) {
       cls = dataProviderClass;
       shouldBeStatic = true;
     }

--- a/testng-core/src/test/java/test/dataprovider/DataProviderTest.java
+++ b/testng-core/src/test/java/test/dataprovider/DataProviderTest.java
@@ -43,6 +43,12 @@ import test.dataprovider.issue2565.SampleTestUsingSupplier;
 
 public class DataProviderTest extends SimpleBaseTest {
 
+  @Test(description = "GITHUB-2800")
+  public void testDataProviderFromAbstractClassWhenCoupledWithFactories() {
+    InvokedMethodNameListener listener = run(test.dataprovider.issue2800.TestClassGenerator.class);
+    assertThat(listener.getSucceedMethodNames()).containsExactly("hi", "hi");
+  }
+
   @Test(description = "GITHUB-1691")
   public void testDataProviderInfoIgnored() {
     InvokedMethodNameListener listener =

--- a/testng-core/src/test/java/test/dataprovider/issue2800/AbstractTestClassGenerator.java
+++ b/testng-core/src/test/java/test/dataprovider/issue2800/AbstractTestClassGenerator.java
@@ -1,0 +1,17 @@
+package test.dataprovider.issue2800;
+
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Factory;
+
+public abstract class AbstractTestClassGenerator {
+
+  @DataProvider(name = "dataProvider")
+  public Object[][] dataProvider() {
+    return new Object[][] {{"foo"}, {"bar"}};
+  }
+
+  @Factory(dataProvider = "dataProvider")
+  public Object[] testFactory(String value) {
+    return new Object[] {new TestClassSample()};
+  }
+}

--- a/testng-core/src/test/java/test/dataprovider/issue2800/TestClassGenerator.java
+++ b/testng-core/src/test/java/test/dataprovider/issue2800/TestClassGenerator.java
@@ -1,0 +1,3 @@
+package test.dataprovider.issue2800;
+
+public class TestClassGenerator extends AbstractTestClassGenerator {}

--- a/testng-core/src/test/java/test/dataprovider/issue2800/TestClassSample.java
+++ b/testng-core/src/test/java/test/dataprovider/issue2800/TestClassSample.java
@@ -1,0 +1,9 @@
+package test.dataprovider.issue2800;
+
+import org.testng.annotations.Test;
+
+public class TestClassSample {
+
+  @Test
+  public void hi() {}
+}

